### PR TITLE
[SPARK-17616][SQL] Support a single distinct aggregate combined with a non-partial aggregate

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

We currently cannot execute an aggregate that contains a single distinct aggregate function and an one or more non-partially plannable aggregate functions, for example:

``` sql
select   grp, 
         collect_list(col1),
         count(distinct col2)
from     tbl_a
group by 1
```

This is a regression from Spark 1.6. This is caused by the fact that the single distinct aggregation code path assumes that all aggregates can be planned in two phases (is partially aggregatable). This PR works around this issue by triggering the `RewriteDistinctAggregates` in such cases (this is similar to the approach taken in 1.6).
## How was this patch tested?

Created `RewriteDistinctAggregatesSuite` which checks if the aggregates with distinct aggregate functions get rewritten into two `Aggregates` and an `Expand`. Added a regression test to `DataFrameAggregateSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @JoshRosen @rxin

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  test("multiple distinct groups") {
+    val input = testRelation
+      .groupBy('a)(countDistinct('b, 'c), countDistinct('d), sum('e))
+      .analyze
+    checkRewrite(RewriteDistinctAggregates(input))
+  }
+
+  test("multiple distinct groups without non-distinct aggregates") {
```
Comment (srinathshankar) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: Do you mean non-partial aggregates here ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    case _ => fail(s"Plan is not rewritten:\n$rewrite")
+  }
+
+  test("single distinct group") {
+    val input = testRelation
+      .groupBy('a)(countDistinct('e))
+      .analyze
+    val rewrite = RewriteDistinctAggregates(input)
+    comparePlans(input, rewrite)
+  }
```
Comment (srinathshankar) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: Could you also add single distinct group with aggregates that have partial

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        CollectSet('b).toAggregateExpression().as('agg2))
+      .analyze
+    checkRewrite(RewriteDistinctAggregates(input))
+  }
+
+  test("multiple distinct groups") {
+    val input = testRelation
+      .groupBy('a)(countDistinct('b, 'c), countDistinct('d), sum('e))
+      .analyze
+    checkRewrite(RewriteDistinctAggregates(input))
```
Comment (srinathshankar) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: Could you also add a test with partials, and one without partials here ? (part of the same test(""))

**Review by srinathshankar — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: I actually mean that the test only contains distinct aggregates.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala (hunk 2 above)
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala (hunk 3 above)
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: done

**Review by srinathshankar — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging to master/2.0. Thanks for the review.

**Pull request closed without merging**